### PR TITLE
support /v1/health using a generation 1 token

### PR DESCRIPTION
**Pull request by LucienShui**
## Motivation

#853

## Modification

Call generate in `/heath`.

## Checklist

- [x] Before submitting a PR for review, make sure it has passed verification in your local development environment **at least**.
- [x] Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
- [ ] Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
- [ ] Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/server.py [hunk 1]
```diff
@@ -90,9 +90,17 @@
 
 
 @app.get("/health")
-async def health() -> Response:
```
Comment (zhyncs) [inline python/sglang/srt/server.py:93]: May we add a `/v1/health` separately?

File: python/sglang/srt/server.py [hunk 2]
```diff
@@ -90,9 +90,17 @@
...
-    """Health check."""
-    return Response(status_code=200)
+async def health(request: Request) -> Response:
+    """
+    Generate 1 token to verify the health of the inference service.
+    """
+    gri = GenerateReqInput(text="s", sampling_params={"max_new_tokens": 1, "temperature": 0.7})
```
Comment (zhyncs) [inline python/sglang/srt/server.py]: Why use temperature 0.7?

File: python/sglang/srt/server.py [hunk 3]
```diff
@@ -90,9 +90,17 @@
...
+    try:
+        _ = await tokenizer_manager.generate_request(gri, request).__anext__()
+        return Response(status_code=200)
+    except Exception as e:
+        logger.exception(e)
+        return Response(status_code=500)
```
Comment (zhyncs) [inline python/sglang/srt/server.py]: Why use status code 500?

File: python/sglang/srt/server.py (hunk 2 above)
Comment (LucienShui) [reply python/sglang/srt/server.py]: Just a random number, I want to verify the sampling at the same time.

File: python/sglang/srt/server.py (hunk 1 above)
Comment (LucienShui) [reply python/sglang/srt/server.py:93]: Good advice!

File: python/sglang/srt/server.py (hunk 3 above)
Comment (LucienShui) [reply python/sglang/srt/server.py]: Good question, maybe 503 is better?

Comment (zhyncs) [reply python/sglang/srt/server.py]: Yes. I think so. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503

**Comment by zhyncs** [comment]
It would be better to write a test case for this URI to verify whether it meets expectations when it is genuinely available and when it is not.

**Comment by LucienShui** [comment]
Honestly, I tried, but I don't know how to simulate a failure that keeps the inference service down without affecting the web service to verify the healthcheck.

I tried causing exceptions manually, but the service quit immediately. Then I reverted the code to v0.2.5, then raise an Exception at the same line as #853, but it still quit immediately.

Maybe I need some help about the unittest.😢

**Comment by zhyncs** [comment]
Maybe you can use `pytest.raises`?

**Comment by vhain** [comment]
I needed this feature. We've been using `POST /v1/completions {"model":" ","prompt":" ","max_tokens":1}` for custom health check endpoint for autoscaler/autohealer at our infra. Once this merges, we can finally and reliably use `GET /health` (or `/v1/health`) for health checking.

One little concern is under high load, `/health` could timeout even if it is actually running. Maybe in the future, we can design sophisticated health check logic without needing to actually run the inference? but I think it's a good start.

**Comment by LucienShui** [comment]
> One little concern is under high load, /health could timeout even if it is actually running.

@vhain Yeah, I have considered this. Maybe for healthcheck (livenessProbe in k8s), we could set a long timeout, and for load banlance ready check (readinessProbe in k8s), we could set a short timeout.

I thought that both healthcheck and ready check timeout should be determined based on various factors such as different model, devices and context lengths, which making it difficult to find a fixed value.

Could you please share some insights or experiences regarding timeout?

**Comment by vhain** [comment]
@LucienShui I think I need your insight as well 😅 I think we can set distinct timeout for each hardware config, and model settings as you mentioned.

In our case, we've been just constantly monitoring the latency for 1 token generation, and found the p95 to be around 5s, so we've set the timeout to 10s. However turned out sometimes it was not enough, so we just bumped up to 30s.

If it's taking more than 30s to generate 1 token, it's bad anyways, so we have the health check to trigger our autoscaler to scale out. If it's timing out for over threshold iterations (currently 2), we mark the instance unhealthy and replace it with the new one.

I think there is no right or wrong answer for LLM backend autoscaling/autohealing at the moment, as things are pretty in early stage.

**Comment by vhain** [comment]
@LucienShui I actually got an idea. Maybe we can add a periodic check (1 token generation; for every 10s or so - configurable by server args), running inside of the SGLang itself, cache the result, and use it for `/health` serving?

This way we can make sure that calling `/health` multiple times does not hurt, also it responds very quickly.

**Comment by LucienShui** [comment]
@vhain I have considered this approach, but it has the following issues:

1. The health check results might represent a historical health status under high load.
2. If let user control the internal loop interval, it would be better to let the user decide the health check interval directly.
3. Additionally, its implementation and runtime complexity are higher.

**Comment by merrymercy** [comment]
Great discussions! Can we merge this first and iterate on better designs in follow-up PRs?
For this PR, please create a separate endpoint and we can merge it.

File: python/sglang/srt/server.py [hunk 4]
```diff
@@ -89,10 +89,34 @@
...
+    gri = GenerateReqInput(text="s", sampling_params={"max_new_tokens": 1, "temperature": 0.7})
+    try:
+        _ = await tokenizer_manager.generate_request(gri, request).__anext__()
+        return Response(status_code=200)
+    except Exception as e:
+        logger.exception(e)
+        return Response(status_code=503)
+
+
 @app.get("/health")
```
Comment (zhyncs) [inline python/sglang/srt/server.py:109]: Don't change `/health`.

**Comment by LucienShui** [comment]
@merrymercy Yeah, sure! I'm glad it can be merged.

I'll trying to add some test cases in another PR.

File: python/sglang/srt/server.py (hunk 4 above)
Comment (LucienShui) [reply python/sglang/srt/server.py:109]: OMG, that's a mistake, I'm sorry.

Comment (LucienShui) [reply python/sglang/srt/server.py:109]: I'm not very familiar with the `rebase` command. Sorry about that again.

File: python/sglang/srt/server.py [hunk 5]
```diff
@@ -89,6 +89,22 @@
...
+@app.get("/v1/health")
+async def health(request: Request) -> Response:
+    """
+    Generate 1 token to verify the health of the inference service.
+    """
+    gri = GenerateReqInput(
+        text="s", sampling_params={"max_new_tokens": 1, "temperature": 0.7}
+    )
+    try:
+        _ = await tokenizer_manager.generate_request(gri, request).__anext__()
```
Comment (zhyncs) [inline python/sglang/srt/server.py]: We should be compatible with Python 3.8.

Comment (zhyncs) [reply python/sglang/srt/server.py]: Although we have manually upgraded Python to 3.10 in the Docker images we released, considering that some users are using Python 3.8 by default on Ubuntu 20.04, the `__anext__` used is only available after Python 3.10.

Comment (zhyncs) [reply python/sglang/srt/server.py]:
```suggestion
        async for _ in tokenizer_manager.generate_request(gri, request):
            break
```

Comment (LucienShui) [reply python/sglang/srt/server.py]: Will using `async for _ in tokenizer_manager.generate_request(gri, request)` solve this issue?

I couldn't find a Python 3.8 sample in our repo, Did I miss something?

Comment (LucienShui) [reply python/sglang/srt/server.py]: Oh wow, that's exactly what I was thinking, lol.

**Review by zhyncs — APPROVED** [review]

**Comment by zhyncs** [comment]
@LucienShui Thanks for your contribution!

**Pull request merged**